src/platform/posix/posix_pollq_kqueue.c

```diff
@@ -0,0 +1,309 @@
+#ifdef NNG_HAVE_KQUEUE
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h> /* for strerror() */
+#include <sys/event.h>
+#include <unistd.h>
+
+#include "core/nng_impl.h"
+#include "platform/posix/posix_pollq.h"
+
+#if defined NNG_PLATFORM_NETBSD
+#define kevent_udata_t intptr_t
+#else
+#define kevent_udata_t void *
+#endif
+
+// nni_posix_pollq is a work structure that manages state for the kqueue-based
+// pollq implementation
+struct nni_posix_pollq {
+	nni_mtx               mtx;
+	nni_cv                cv;
+	int                   kq;      // kqueue handle
+	int                   wakewfd; // write side of waker pipe
+	int                   wakerfd; // read side of waker pipe
+	int                   close;   // request for worker to exit
+	int                   started;
+	nni_thr               thr;    // worker thread
+	nni_posix_pollq_node *wait;   // cancel waiting on this
+	nni_posix_pollq_node *active; // active node (in callback)
+};
+
+int
+nni_posix_pollq_add(nni_posix_pollq *pq, nni_posix_pollq_node *node)
+{
+	if (pq == NULL) {
+		return (NNG_EINVAL);
+	}
+	if (pq->close) {
+		// This shouldn't happen!
+		nni_mtx_unlock(&pq->mtx);
+		return (NNG_ECLOSED);
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	node->pq = pq;
+	nni_mtx_unlock(&pq->mtx);
+
+	return (0);
+}
+
+void
+nni_posix_pollq_remove(nni_posix_pollq_node *node)
+{
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	while (pq->active == node) {
+		pq->wait = node;
+		nni_cv_wait(&pq->cv);
+	}
+
+	if (pq->close) {
+		nni_cv_wake(&pq->cv);
+	}
+	node->pq = NULL;
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_arm(nni_posix_pollq_node *node, int events)
+{
+	nni_posix_pollq *pq = node->pq;
+	struct kevent    kevents[2];
+	int              nevents = 0;
+
+	if (pq == NULL || events == 0) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	if (!(node->events & POLLIN) && (events & POLLIN)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_READ, EV_ADD, 0,
+		    0, (kevent_udata_t) node);
+	}
+
+	if (!(node->events & POLLOUT) && (events & POLLOUT)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_WRITE, EV_ADD, 0,
+		    0, (kevent_udata_t) node);
+	}
+
+	if (nevents > 0) {
+		int rv;
+		rv = kevent(pq->kq, kevents, nevents, NULL, 0, NULL);
+		if (rv < 0) {
```

Comment:
Looking at this, I'm inclined to think what we ought to do is mark the event as if POLLERR was triggered.  It would be even handier if we could stash the errno somewhere, or some indication that it was *kevent* that triggered the error.  Stay tuned.